caravel/models.py

```diff
@@ -370,6 +441,100 @@ def json_data(self):
+    def import_dashboard(cls, dashboard_to_import, import_time=None):
+        """Imports the dashboard from the object to the database.
+
+        Once dashboard is imported, json_metadata field is extended and stores
+        remote_id and import_time. It helps to decide if the dashboard has to
+        be overridden or just copies over. Slices that belong to this dashboard
+        will be wired to existing tables using Slice.perm field to derive the
+        datasource name. This function can be used to import/export dashboards
+        between multiple caravel instances. Audit metadata isn't copies over.
+        """
+        logging.info('Started import of the dashboard: {}'
+                     .format(dashboard_to_import.to_json()))
+        make_transient(dashboard_to_import)
+        # TODO: provide nice error message about failed imports
+        slices_to_attach = []
+        session = db.session
+        logging.info('Dashboard has {} slices'
+                     .format(len(dashboard_to_import.slices)))
+        # copy slices object as Slice.import_slice will mutate the slice
+        # and will remove the existing dashboard - slice association
+        slices = copy(dashboard_to_import.slices)
+        for slc in slices:
+            logging.info('Importing slice {} from the dashboard: {}'.format(
+                slc.to_json(), dashboard_to_import.dashboard_title))
+            slc_id = Slice.import_slice(slc, import_time=import_time)
+            if not slc_id:
+                continue
+            slices_to_attach.append(
+                session.query(Slice).filter_by(id=slc_id).first())
+            logging.info(
+                slc.slice_name + ' belongs to the dashboard ' +
+                dashboard_to_import.dashboard_title)
+        # Skip dashboard creation if the slice wasn't imported
+        if not slices_to_attach:
+            logging.warning('Dashboard {} import failed'.format(
+                dashboard_to_import.dashboard_title))
+            return
+
+        if dashboard_to_import.json_metadata:
+            json_metadata_dict = json.loads(dashboard_to_import.json_metadata)
+        else:
+            json_metadata_dict = {}
+        if 'remote_id' not in json_metadata_dict:
+            json_metadata_dict['remote_id'] = dashboard_to_import.id
+            json_metadata_dict['import_time'] = import_time
+
+        # Override the dashboard
+        existing_dashboard = (
+            session.query(Dashboard)
+            .filter(Dashboard.json_metadata.ilike(
+                '%"remote_id": {}%'.format(dashboard_to_import.id)))
+            .first()
+        )
+        dashboard_to_import.id = None
+        if existing_dashboard:
+            # TODO: add nice notification about the overriden dashboard
+            logging.info('Dashboard already exist and will be overridden: {}'
+                         .format(existing_dashboard.to_json()))
+
+            session.delete(existing_dashboard)
```

Comment:
deleting means the id will change and that bookmarked urls won't work, let's just merge instead of deleting / creating